From dev to main (week #3)

ZhosanDenis opened the pull request.

ZhosanDenis committed:
Изменения №1 от 16.06.2023
- исправления ошибок для корректной работы с сайтом

ZhosanDenis committed:
Изменения №2 от 16.06.2023
- дополнительные исправления ошибок для корректной работы с сайтом
- добавление эндпоинта для поиска объявления по названию

ZhosanDenis committed:
Merge pull request #13 from Zhoss/1-feature-denis

Изменения №1 от 16.06.2023

ZhosanDenis committed:
Изменения №1 от 18.06.2023
- AccountServiceImpl: функционал по нахождению и загрузке изображения выведен в отдельный метод для использования в AdServiceImpl - downloadAdImageFromFS();
- AdController: добавлен эндпоинт для загрузки изображения объявления со стороны фронта;
- AdService и AdServiceImpl: добавлен функционал (метод) для загрузки изображения объявления;
- в импортах сущностей удалены .* (заменены на необходимые импорты).

ZhosanDenis committed:
Merge branch 'dev' into 1-feature-denis

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdController.java:
```diff
@@ -251,4 +243,13 @@ public ResponseEntity<?> deleteComment(
+            summary = "Найти объявление по названию",
+            tags = "Объявления"
+    )
+    @GetMapping("/find/{title}")
```
title лучше сделать параметром запроса, а не пути. А вообще, можно было бы просто сделать его опциональным параметром запроса для /ads, чтобы не создавать новый эндпоинт

ZhosanDenis replied:
Опциональный параметр запроса - через required = false?

dmitry-bizin replied:
ага

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AccountServiceImpl.java:
```diff
@@ -33,25 +34,27 @@ public class AccountServiceImpl implements AccountService {
     public boolean updatePassword(NewPassword newPassword, String userName) {
-        if (userRepository.existsByEmail(userName) &&
-                userRepository.existsByPassword(Objects.hash(userName, newPassword.getCurrentPassword()))) {
+        if (userRepository.existsByPassword(Math.abs(Objects.hash(userName, newPassword.getCurrentPassword())))) {
```
ой, а зачем вы так накрутили проверку-то?) у вас и так будет хэшироваться пароль дефолтным PasswordEncoder-ом, вам не нужно брать от него ещё хэшкод и проверять его, тем более модуль этого числа. И к тому же, почему проверяется хэшкод юзера и пароля??) надо проверять хэши паролей - старого из БД и старого из NewPassword, эти пароли надо сравнивать

ZhosanDenis replied:
Дмитрий, проверка user.getPassword().equals(encoder.encode(newPassword.getCurrentPassword())) возвращает false и пароль не меняется. Посмотрел - уже захешированный пароль из БД и хэшкод текущего пароля отличаются. Т.е. encoder один и тот же пароль хэширует по-разному? В чем ошибка?

dmitry-bizin replied:
Да, для bcrypt хэши могут быть разными, поэтому надо проверять через метод encoder.matches, который принимает первым параметром пароль в открытом виде (из запроса), а вторым - хэшированный из БД

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AccountServiceImpl.java:
```diff
@@ -33,25 +34,27 @@ public class AccountServiceImpl implements AccountService {
+        if (userRepository.existsByPassword(Math.abs(Objects.hash(userName, newPassword.getCurrentPassword())))) {
             UserEntity user = userRepository.findByEmail(userName);
-            user.setPassword(Objects.hash(userName, newPassword.getNewPassword()));
+            user.setPassword(Math.abs(Objects.hash(userName, newPassword.getNewPassword())));
```
здесь надо просто использовать PasswordEncoder (просто заинжектите его), чтобы захэшировать пароль

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/UserMapperImpl.java:
```diff
@@ -14,7 +14,7 @@ public class UserMapperImpl implements UserMapper {
         UserEntity userEntity = new UserEntity();
         userEntity.setEmail(req.getUsername());
-        userEntity.setPassword(Objects.hash(req.getUsername(), req.getPassword()));
+        userEntity.setPassword(Math.abs(Objects.hash(req.getUsername(), req.getPassword())));
```
здесь тоже надо через PasswordEncoder

ZhosanDenis committed:
Изменения №1 от 19.06.2023
- изменен тип пароля для хранения в БД (с long на String);
- изменен способ хэширования пароля (с помощью PasswordEncoder) при создании пользователя и обновлении пароля;
- изменена логика проверки пароля (текущего от фронта и в БД) при его обновлении.

ZhosanDenis committed:
Merge pull request #15 from Zhoss/1-feature-denis

Изменения №1 от 18.06.2023
